### PR TITLE
Fix several GMap issues

### LINKED ISSUE #2822: GMapPlot Resets to (0, 0)
ResetTool() for GmapPlot centers on lat, lon = (0,0), rather than the lat, lon defined in GMapOptions.  

```
from __future__ import print_function
from bokeh.browserlib import view
from bokeh.document import Document
from bokeh.embed import file_html
from bokeh.models.glyphs import Circle
from bokeh.models import (
    GMapPlot, GeoJSPlot, Range1d, ColumnDataSource, LinearAxis,
    PanTool, WheelZoomTool, BoxSelectTool, ResetTool,
    BoxSelectionOverlay, GMapOptions, GeoJSOptions, GeoJSPlot, 
    NumeralTickFormatter, PrintfTickFormatter)
from bokeh.resources import INLINE

gmap_options = GMapOptions(lat=30.2861, lng=-97.7394, map_type="roadmap", zoom=13)

x_range = Range1d()
y_range = Range1d()

plot = GMapPlot(
        x_range=x_range, y_range=y_range,
        map_options=gmap_options,
        title = "Austin")


source = ColumnDataSource(
    data=dict(
        lat=[30.2861, 30.2855, 30.2869],
        lon=[-97.7394, -97.7390, -97.7405],
        fill=['orange', 'blue', 'green']
    )
)

circle = Circle(x="lon", y="lat", size=15, fill_color="fill", line_color="black")
plot.add_glyph(source, circle)

plot.add_tools(PanTool(), WheelZoomTool(), ResetTool())

xaxis = LinearAxis(axis_label="lat", major_tick_in=0, formatter=NumeralTickFormatter(format="0.000"))
plot.add_layout(xaxis, 'below')

yaxis = LinearAxis(axis_label="lon", major_tick_in=0, formatter=PrintfTickFormatter(format="%.3f"))
plot.add_layout(yaxis, 'left')

doc = Document()
doc.add(plot)

if __name__ == "__main__":
    filename = "maps.html"
    with open(filename, "w") as f:
        f.write(file_html(doc, INLINE, "Maps Example"))
    print("Wrote %s" % filename)
    view(filename)
```

### LINKED ISSUE #3737: GMapPlot doesn't display anything in Jupyter v.4
Hi,

Just started using Bokeh more extensively and I wanted to display maps within a Jupyter notebook. I tried running the example provided in the repo (with the world_cities dataset) and while it displays when ran alone, it only displays a blank area in Jupyter (and I could only see the data once (without a map) when I refreshed, but this is not consistent). I have both latest versions for bokeh and Jupyter (from Anaconda). 

```
import bokeh
import bokeh.plotting as plotting
plotting.output_notebook()
from bokeh.io import push_notebook

from bokeh.resources import INLINE
from bokeh.models.glyphs import Circle
from bokeh.models import (
    GMapPlot, Range1d, ColumnDataSource,
    PanTool, WheelZoomTool, GMapOptions)
from bokeh.resources import INLINE
from bokeh.sampledata.world_cities import data


x_range = Range1d(-160, 160)
y_range = Range1d(-80, 80)

map_options = GMapOptions(lat=15, lng=0, zoom=10)

plot = GMapPlot(
    x_range=x_range,
    y_range=y_range,
    plot_width=1000,
    plot_height=500,
    map_options=map_options,
    title="Cities of the world with a population over 5,000 people.",
    webgl=True,
)
circle = Circle(x="lng", y="lat", size=5, line_color=None, fill_color='firebrick', fill_alpha=0.2)
plot.add_glyph(ColumnDataSource(data), circle)
plot.add_tools(PanTool(), WheelZoomTool())
plotting.show(plot)
```

### LINKED ISSUE #4835: Multiple GMap plots whiting out and multiple calls of gmap plot whiteing out in notebook.
Hello, 

I am trying to build interactive bokeh plots using overlayed Google Maps. I am hoping to have similar functionality to regular bokeh plots, being able to call them multiple times in a notebook without issue. However, for Google Maps this doesn't yet seem to be the case. Exactly the first time I call exactly the first map in the following gist the map loads and works. All other maps display first my overlaid points, then pure whiteness. 

This is in Bokeh 0.12.0

https://gist.github.com/cenveoanalyst/8931355cd8ee7466808e0cc452666246

The issue tracker isn't the place for general support. For questions and
technical assistance, come ask the [Bokeh mailing list](https://groups.google.com/a/continuum.io/forum/#!forum/bokeh) or join the chat on [Gitter](https://gitter.im/bokeh/bokeh).

For defects or deficiencies, please provide:
- Description of expected behavior and the observed behavior
- Self-contained example code that generates the issue (if possible)
- Stack traceback and/or browser developer tools console output
- Software version information output by `bokeh info` command

For desired features, please provide:
- Description of the desired feature

### LINKED ISSUE #5442: Add scale control to GMaps
Currently there are many configurable options that are not exposed in `GMapOptions`, such as the ability to add a scale bar. 

The various GMap options are described here:

https://developers.google.com/maps/documentation/javascript/controls

ref: http://stackoverflow.com/questions/40423735/python-bokeh-google-maps-plot-show-1km-distance-based-on-zoom-on-bottom-right

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #5442 fixes #2822 fixes #4835 fixes #3737

This PR:

* fixes reload issue. Was building map too soon, before div dims were set. Moved call to build map to `_map_hook`. Allowed simplifying API load logic, callback is now just a call to request render

* makes `MapOptions` a proper `Model` and wires up missing plumbing to respond to map options properties changes

* stores initial center and zoom correctly so that reset tool functions properly

There is more to fix, but that can happen in follow on PRs

### CODE REVIEW

**Comment by bryevdv** [comment]
@mattpap I originally tried to be more narrow and add a change handler like:
```
@listenTo(@model.plot.map_options, 'change:styles', () => @_update_options())
```
But this results in an infinite recursions / call stack exceeded error. I am guessing offhand it has to do with `MapOptions` being a `HasProps` and not a `Model`.  Is this supposed to work? Is there any reason not to just make it a `Model` in any case?

**Comment by mattpap** [comment]
@bryevdv, no idea. I may need to have a closer look at this.

**Comment by mattpap** [comment]
I wanted to suggest to add this example as integration test, but I doubt it will work due to maps' load delay.

**Comment by bryevdv** [comment]
@mattpap I'm inclined to try just making `MapOptions` a model as an immediate workaround, and also to add a commented out test for adding listeners to `HasProps` (and a separate issue for addressing that)

**Comment by bryevdv** [comment]
Presumably the reason for not making them Models was to avoid having to make corresponding models on the JS side. I will try fixing up `HasProps` events first.

**Comment by bryevdv** [comment]
Actually I think it does not make sense to add events on `HasProps` and there is another issue to make `listenTo`, etc be strict. I think that means only works on models with properties. So I guess we have to bite the bullet and make models for all the map options objects. That will give fine grained control over setting all the fields in it.

**Comment by bryevdv** [comment]
OK it's working much better but there are several other GMap issues, around panning, range setting, repeated maps not showing... going to try to clean up GMap comprehensively in this PR

**Comment by bryevdv** [comment]
Simple test script:
```
from bokeh.io import curdoc
from bokeh.models.glyphs import Circle
from bokeh.layouts import column
from bokeh.models import GMapPlot, Range1d, ColumnDataSource, PanTool, WheelZoomTool, BoxSelectTool, GMapOptions, Button, ResetTool

x_range = Range1d()
y_range = Range1d()

# JSON style string taken from: https://snazzymaps.com/style/1/pale-dawn
map_options = GMapOptions(lat=30.2861, lng=-97.7394, map_type="roadmap", zoom=13, styles="""
[{"featureType":"administrative","elementType":"all","stylers":[{"visibility":"on"},{"lightness":33}]},{"featureType":"landscape","elementType":"all","stylers":[{"color":"#f2e5d4"}]},{"featureType":"poi.park","elementType":"geometry","stylers":[{"color":"#c5dac6"}]},{"featureType":"poi.park","elementType":"labels","stylers":[{"visibility":"on"},{"lightness":20}]},{"featureType":"road","elementType":"all","stylers":[{"lightness":20}]},{"featureType":"road.highway","elementType":"geometry","stylers":[{"color":"#c5c6c6"}]},{"featureType":"road.arterial","elementType":"geometry","stylers":[{"color":"#e4d7c6"}]},{"featureType":"road.local","elementType":"geometry","stylers":[{"color":"#fbfaf7"}]},{"featureType":"water","elementType":"all","stylers":[{"visibility":"on"},{"color":"#acbcc9"}]}]
""")

API_KEY = "<API KEY>"

plot = GMapPlot(
    x_range=x_range, y_range=y_range,
    map_options=map_options,
    api_key=API_KEY,
)
plot.title.text = "Austin"

source = ColumnDataSource(
    data=dict(
        lat=[30.2861, 30.2855, 30.2869],
        lon=[-97.7394, -97.7390, -97.7405],
        fill=['orange', 'blue', 'green']
    )
)

circle = Circle(x="lon", y="lat", size=15, fill_color="fill", line_color="black")
plot.add_glyph(source, circle)

pan = PanTool()
wheel_zoom = WheelZoomTool()
box_select = BoxSelectTool()
reset_tool = ResetTool()

plot.add_tools(pan, wheel_zoom, box_select, reset_tool)

def callback():
    #map_options = GMapOptions(lat=30.2821, lng=-97.7414, map_type="roadmap", zoom=13)
    plot.map_options.styles="""[ { "elementType": "labels", "stylers": [ { "visibility": "off" }, { "color": "#f49f53" } ] }, { "featureType": "landscape", "stylers": [ { "color": "#f9ddc5" }, { "lightness": -7 } ] }, { "featureType": "road", "stylers": [ { "color": "#813033" }, { "lightness": 43 } ] }, { "featureType": "poi.business", "stylers": [ { "color": "#645c20" }, { "lightness": 38 } ] }, { "featureType": "water", "stylers": [ { "color": "#1994bf" }, { "saturation": -69 }, { "gamma": 0.99 }, { "lightness": 43 } ] }, { "featureType": "road.local", "elementType": "geometry.fill", "stylers": [ { "color": "#f19f53" }, { "weight": 1.3 }, { "visibility": "on" }, { "lightness": 16 } ] }, { "featureType": "poi.business" }, { "featureType": "poi.park", "stylers": [ { "color": "#645c20" }, { "lightness": 39 } ] }, { "featureType": "poi.school", "stylers": [ { "color": "#a95521" }, { "lightness": 35 } ] }, {}, { "featureType": "poi.medical", "elementType": "geometry.fill", "stylers": [ { "color": "#813033" }, { "lightness": 38 }, { "visibility": "off" } ] }, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, { "elementType": "labels" }, { "featureType": "poi.sports_complex", "stylers": [ { "color": "#9e5916" }, { "lightness": 32 } ] }, {}, { "featureType": "poi.government", "stylers": [ { "color": "#9e5916" }, { "lightness": 46 } ] }, { "featureType": "transit.station", "stylers": [ { "visibility": "off" } ] }, { "featureType": "transit.line", "stylers": [ { "color": "#813033" }, { "lightness": 22 } ] }, { "featureType": "transit", "stylers": [ { "lightness": 38 } ] }, { "featureType": "road.local", "elementType": "geometry.stroke", "stylers": [ { "color": "#f19f53" }, { "lightness": -10 } ] }, {}, {}, {} ] """
    plot.map_options.lat += 0.01
    plot.map_options.lng += 0.01
    #plot.map_options = map_options

button = Button(label="Foo")
button.on_click(callback)

curdoc().add_root(column(button, plot))
```

**Comment by bryevdv** [comment]
@mattpap do you have an comments? I'd like to merge this and get a dev build out ASAP so that some people can test these GMap fixes as well as the one from the ML today.

File: bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee [hunk 1]
```diff
@@ -56,86 +57,115 @@ export class GMapPlotCanvasView extends PlotCanvasView
...
+    @map_types = {
+      satellite : maps.MapTypeId.SATELLITE,
+      terrain   : maps.MapTypeId.TERRAIN,
+      roadmap   : maps.MapTypeId.ROADMAP,
+      hybrid    : maps.MapTypeId.HYBRID
+    }
 
-    @initial_zoom = @model.plot.map_options.zoom
+    mo = @model.plot.map_options
+    map_options =
```
Comment (mattpap) [inline bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee]: Make sure to not use implicit object and function call literals in new code.

File: bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee [hunk 2]
```diff
@@ -4,44 +4,45 @@ import {PlotCanvas, PlotCanvasView} from "./plot_canvas"
 import * as p from "../../core/properties"
 import {defer} from "../../core/util/callback"
 
+load_google_api = (callback, api_key) ->
+  if not window.google?.maps?
+    window._bokeh_gmap_callback = () -> callback()
```
Comment (mattpap) [inline bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee]: This prevents loading two versions of bokeh on one page.

**Comment by mattpap** [comment]
Yeah, I have some comments, but nothing that has to be resolved here.

**Comment by mattpap** [comment]
GitHub's UI is just brilliant. I almost closed this PR again when writing this comment.

File: bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee (hunk 1 above)
Comment (bryevdv) [inline bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee]: I'l push a fix for this before merging

File: bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee (hunk 2 above)
Comment (bryevdv) [inline bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee]: I can restore the more complicated "list of callbacks" if you think it's a priority.

Comment (mattpap) [inline bokehjs/src/coffee/models/plots/gmap_plot_canvas.coffee]: As I said in a top-level comment, both issues are low-priority.

**Pull request merged**